external/radiation/radiation/preprocessing.py

```diff
@@ -0,0 +1,257 @@
+import numpy as np
+import xarray as xr
+import cftime
+
+try:
+    from vcm.calc.thermo.vertically_dependent import (
+        pressure_at_interface,
+        pressure_at_midpoint_log,
+    )
+    from vcm.calc.thermo.constants import _SPECIFIC_HEAT_CONST_PRESSURE, _RDGAS
+except ImportError:
+    pass
+
+P_REF = 1.0e5
+MINUTES_PER_HOUR = 60.0
+SECONDS_PER_MINUTE = 60.0
+TRACER_NAMES_IN_MAPPING: Mapping[str, str] = {  # this is specific to GFS physics
+    "cloud_water_mixing_ratio": "ntcw",
+    "rain_mixing_ratio": "ntrw",
+    "cloud_ice_mixing_ratio": "ntiw",
+    "snow_mixing_ratio": "ntsw",
+    "graupel_mixing_ratio": "ntgl",
+    "ozone_mixing_ratio": "ntoz",
+    "cloud_amount": "ntclamt",
+}
+SFC_NAMES_IN_MAPPING: Mapping[str, str] = {
+    "surface_temperature": "tsfc",
+    "land_sea_mask": "slmsk",
+    "snow_depth_water_equivalent": "snowd",
+    "snow_cover_in_fraction": "sncovr",
+    "maximum_snow_albedo_in_fraction": "snoalb",
+    "surface_roughness": "zorl",
+    "orographic_variables": "hprime",
+    "mean_visible_albedo_with_strong_cosz_dependency": "alvsf",
+    "mean_near_infrared_albedo_with_strong_cosz_dependency": "alnsf",
+    "mean_visible_albedo_with_weak_cosz_dependency": "alvwf",
+    "mean_near_infrared_albedo_with_weak_cosz_dependency": "alnwf",
+    "fractional_coverage_with_strong_cosz_dependency": "facsf",
+    "fractional_coverage_with_weak_cosz_dependency": "facwf",
+    "ice_fraction_over_open_water": "fice",
+    "surface_temperature_over_ice_fraction": "tisfc",
+}
+RENAME_OUT: Mapping[str, str] = {
+    "sfcflw_dnfx0": "clear_sky_downward_longwave_flux_at_surface_python",
+    "sfcfsw_dnfx0": "clear_sky_downward_shortwave_flux_at_surface_python",
+    "sfcflw_upfx0": "clear_sky_upward_longwave_flux_at_surface_python",
+    "sfcfsw_upfx0": "clear_sky_upward_shortwave_flux_at_surface_python",
+    "topflw_upfx0": "clear_sky_upward_longwave_flux_at_top_of_atmosphere_python",
+    "topfsw_upfx0": "clear_sky_upward_shortwave_flux_at_top_of_atmosphere_python",
+    "sfcflw_dnfxc": "total_sky_downward_longwave_flux_at_surface_python",
+    "sfcfsw_dnfxc": "total_sky_downward_shortwave_flux_at_surface_python",
+    "sfcflw_upfxc": "total_sky_upward_longwave_flux_at_surface_python",
+    "sfcfsw_upfxc": "total_sky_upward_shortwave_flux_at_surface_python",
+    "topfsw_dnfxc": "total_sky_downward_shortwave_flux_at_top_of_atmosphere_python",
+    "topflw_upfxc": "total_sky_upward_longwave_flux_at_top_of_atmosphere_python",
+    "topfsw_upfxc": "total_sky_upward_shortwave_flux_at_top_of_atmosphere_python",
+    "htrlw": "total_sky_longwave_heating_rate_python",
+    "lwhc": "clear_sky_longwave_heating_rate_python",
+    "htrsw": "total_sky_shortwave_heating_rate_python",
+    "swhc": "clear_sky_shortwave_heating_rate_python",
+}
+
+State = MutableMapping[Hashable, xr.DataArray]
+Diagnostics = MutableMapping[Hashable, xr.DataArray]
+
+
+def model(
+    rad_config: MutableMapping[Hashable, Any],
```

Comment:
Does it need to mutable? Can you make it a TypeDict or dataclass...which would be valuable documentation.

Reply:
Dpme